### ISSUE TITLE
Bazel doesn't cleanup partial input downloads for remote_download_minimal

### ISSUE THREAD

**Issue by sohaibiftikhar** [bazelbuild/bazel#11339]
### Description of the problem / feature request:

The build without the bytes feature downloads remotely located output files locally only if needed to execute an action locally. This is done as follows:
1. It downloads the file at the exact output path where the actual action would have generated it hence making it transparent to the local action where the previous action was executed. 
2. It [marks the permissions](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteActionInputFetcher.java;l=189;drc=f17c403b2365ff0268b0ecae2e00e95fcad9d36b?q=RemoteActionInputFetcher&ss=bazel) of the local file to 0755.
3. On completion of the action, the input files are [removed from the local disk](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java;drc=f17c403b2365ff0268b0ecae2e00e95fcad9d36b;l=600?q=RemoteModule&ss=bazel) to make the local action cache in sync with the local output base.

However, since Bazel executes actions in parallel if any other build action fails this cancels all other concurrent actions using the thread executor service (I assume). This can happen at multiple points:
<a name="scenario1"></a>
1. During/Before (1) above. This leads to an incomplete download. On the next run bazel sees that the digest of the output does not match the remotely expected digest and hence reruns the action. This is okay.
<a name="scenario2"></a>
2. After (1) above. Since the file is completely downloaded and the digest does not take permissions into account, Bazel assumes the file is completely downloaded and attempts to execute the action using the file as an input. **However, this is not okay.** Some actions like `ExecutableSymlink` expect the file to be executable before the action is executed on it and can fail.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.
<a name="affected_versions"></a>
The bug is very hard to reproduce. There are a few relatively easier ways to reproduce it.
1. Take bazel_version < 1.2.0.
2. Revert the [context propagation to the Grpc Server](https://github.com/bazelbuild/bazel/commit/fdfd88d3e9885f4f28d1ce8f2cee4461d6cb0ef4) in bazel version after 1.2.0.
3. Take a commit from the master after [this commit](https://github.com/bazelbuild/bazel/commit/3ef8fb98e7becaac160003f5c24f35914b764bb9). Because it seems to have broken context propagation once again.
**Note:** In all these cases failed builds will happen even with [scenario 1](#scenario1) mentioned above since the download continues in the background after build-failure. (This seems to be a separate issue).

Steps (needs a remote-cache):
1. Create a minimal BUILD file
```
genrule(
     name = "big",
     outs = ["test.sh"],
     # This is just to create a large enough executable that needs to be cached
     cmd = "(for i in `seq 1 1000000`; do echo '#some script'; done && echo 'echo done.') > $@",
     executable = 1,
)
sh_binary(
     name = "big_link",
     srcs = ["big"],
     data = [":big"],
)
```
2. Now run:
```
bazel build --remote_cache="grpc://my-cache" //...
```
3. To emulate another machine that does not have the outputs from this run a `bazel clean`.
4. Now add a failing action to the `BUILD` file. The sleep gives enough time for the download to at least start.
```
genrule(
    name = "genrule",
    outs = ["test.txt"],
    cmd = "sleep 1 && exit 1 && echo 'TEST' > $@",
)
```
Run 
```
bazel build --remote_download_minimal --remote_cache="grpc://my-cache" //...
```
Bazel would fail with something like `non zero exit code from 'genrule'`. If you look inside `bazel-bin` you can find a partially downloaded output for the action `big`. If you used [one of the methods mentioned above for the version of Bazel](#affected_versions) the downloaded file's size would be the same as the expected size but the permissions wouldn't be set. (no context propagation means the download thread continues even after the build fails hence the download itself completes).
5. Now fix the failing action by removing the `exit 1` to change `genrule` to:
```
genrule(
    name = "genrule",
    outs = ["test.txt"],
    cmd = "sleep 1 && echo 'HELLO' > $@",
)
```
Run
```
bazel build --remote_download_minimal --remote_cache="grpc://my-cache" //...
```
Bazel would fail with a message like:
```
ERROR: /home/iftikhso/code/binary_test/big/src/BUILD:40:10: failed to create symbolic link 'src/big_link': file 'test.sh' is not executable
```

### What operating system are you running Bazel on?

Ubuntu 18

### What's the output of `bazel info release`?

1.2.1 but also affects master.

### If `bazel info release` returns "development version" or "(@non-git)", tell us how you built Bazel.

NA

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

NA

###  Have you found anything relevant by searching the web?

No

### Any other information, logs, or outputs that you want to share?

A simple fix might to use the same approach as used by the RemoteCache. It [downloads files to a temporary location](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/RemoteCache.java;l=310;drc=f17c403b2365ff0268b0ecae2e00e95fcad9d36b?q=RemoteCache&ss=bazel) before moving them to their final location.

Even though this bug is somewhat hard to reproduce for versions `1.2.0>=v<=3.1.0)` this is not merely a theoretical issue, and our CI is running into this regularly, and the result is that the local state is corrupt which will not self-heal unless a `bazel clean` is issued.

**Comment by aiuto** [comment]
Is this reproducible in a current Bazel? It sounds like you have to make  a very specific old configuration to make this happen.

**Comment by sohaibiftikhar** [comment]
Its reproducible in the current master.
The problem also exists in 3.1 but is much harder to reproduce there.

**Comment by coeuvre** [comment]
The root cause is that the download future doesn't get cancelled on `InterruptedException`. Fixed by #12453

**Comment by coeuvre** [comment]
Now, if other concurrent actions failed, or entire build is canceled, the downloading process will be stopped and the partial downloaded files will be deleted.

**Comment by sohaibiftikhar** [comment]
Cool. Glad to see this fixed! There is still a race though. What if download just finished and the path was not marked executable yet?
So I think we should still do this.

**Comment by coeuvre** [comment]
Yes, let's track this issue here. #12454
